src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -374,7 +375,44 @@ private class RsFnInferenceContext(
     }
 
     private fun inferPathExprType(expr: RsPathExpr): Ty {
-        val (element, subst) = expr.path.reference.advancedResolve()?.downcast<RsNamedElement>() ?: return TyUnknown
+        val variants = expr.path.reference.advancedCachedMultiResolve().mapNotNull { it.downcast<RsNamedElement>() }
+        val qualifier = expr.path.path
+        if (variants.size > 1 && qualifier != null) {
+            val resolved = resolveAmbiguity(variants)
+            if (resolved != null) {
+                val self = qualifier.reference.advancedResolve()?.downcast<RsNamedElement>()
+                if (self != null) {
+                    // This works for `String::from` where `selfTy` is `String` (in opposition to From::from)
+                    val selfTy = instantiatePath(self)
+                    return instantiatePath(resolved, selfTy)
+                }
+            }
+        }
+        return instantiatePath(variants.firstOrNull() ?: return TyUnknown)
+    }
+
+    /** This works for `String::from` where impls of `From` trait found for `String` */
+    private fun resolveAmbiguity(elements: List<BoundElement<RsNamedElement>>): BoundElement<RsFunction>? {
+        if (elements.size <= 1) return null
+
+        val traits = elements.asSequence()
+            .mapNotNull { it.element as? RsFunction }
+            .mapNotNull { it.owner as? RsFunctionOwner.Impl }
+            .mapNotNull { it.impl.traitRef?.resolveToTrait }
```

Comment:
Perhaps it's better to fuse these mapNotNull calls together?

Reply:
Heh, looks awful
```kotlin
val traits = elements.map {
            ((it.element as? RsFunction)?.owner as? RsFunctionOwner.Impl)?.impl?.traitRef?.resolveToTrait
        }
```